From 4482decd36f9af955f232075b1b2c7627ad5e7c8 Mon Sep 17 00:00:00 2001
From: LR-Vensuki <260785366+LR-Vensuki@users.noreply.github.com>
Date: Wed, 7 Oct 2026 23:12:41 +0300
Subject: [PATCH 03/42] SpiderMonkey on armv7 iOS: pthread mutex for
 mozjemalloc, VFP state for wasm traps, malloc alignment instead of
 max_align_t

---
 js/public/Utility.h                | 17 +++++++++++++++--
 js/src/vm/ArrayBufferObject.h      |  2 +-
 js/src/vm/TypedArrayObject-inl.h   |  2 +-
 js/src/wasm/WasmSignalHandlers.cpp |  7 ++++---
 memory/build/Mutex.cpp             |  2 +-
 memory/build/Mutex.h               | 22 ++++++++++++++--------
 6 files changed, 36 insertions(+), 16 deletions(-)

diff --git a/js/public/Utility.h b/js/public/Utility.h
index 0c9ff326e3..c1e66a2929 100644
--- a/js/public/Utility.h
+++ b/js/public/Utility.h
@@ -16,6 +16,19 @@
 #include <string.h>
 #include <type_traits>
 #include <utility>
+#include <cstddef>
+
+/*
+ * The alignment malloc (and mozjemalloc) guarantees. Everywhere but armv7
+ * Darwin that is max_align_t's; there the APCS aligns 8-byte types to 4, so
+ * max_align_t says 4, while Darwin's malloc and mozjemalloc both return at
+ * least 8-byte aligned blocks for anything 8 bytes or larger.
+ */
+#if defined(__APPLE__) && defined(__arm__)
+#  define JS_MALLOC_ALIGNMENT size_t(8)
+#else
+#  define JS_MALLOC_ALIGNMENT alignof(std::max_align_t)
+#endif
 
 #include "jstypes.h"
 #include "mozmemory.h"
@@ -497,7 +510,7 @@ static inline void js_free(void* p) {
   template <class T, typename... Args>                                      \
   QUALIFIERS T* MOZ_HEAP_ALLOCATOR NEWNAME(Args&&... args) {                \
     static_assert(                                                          \
-        alignof(T) <= alignof(max_align_t),                                 \
+        alignof(T) <= JS_MALLOC_ALIGNMENT,                                 \
         "over-aligned type is not supported by JS_DECLARE_NEW_METHODS");    \
     void* memory = ALLOCATOR(sizeof(T));                                    \
     return MOZ_LIKELY(memory) ? new (memory) T(std::forward<Args>(args)...) \
@@ -516,7 +529,7 @@ static inline void js_free(void* p) {
   template <class T, typename... Args>                                         \
   QUALIFIERS T* MOZ_HEAP_ALLOCATOR NEWNAME(arena_id_t arena, Args&&... args) { \
     static_assert(                                                             \
-        alignof(T) <= alignof(max_align_t),                                    \
+        alignof(T) <= JS_MALLOC_ALIGNMENT,                                    \
         "over-aligned type is not supported by JS_DECLARE_NEW_ARENA_METHODS"); \
     void* memory = ALLOCATOR(arena, sizeof(T));                                \
     return MOZ_LIKELY(memory) ? new (memory) T(std::forward<Args>(args)...)    \
diff --git a/js/src/vm/ArrayBufferObject.h b/js/src/vm/ArrayBufferObject.h
index 5aba8c2807..b55fd2f90b 100644
--- a/js/src/vm/ArrayBufferObject.h
+++ b/js/src/vm/ArrayBufferObject.h
@@ -376,7 +376,7 @@ class ArrayBufferObject : public ArrayBufferObjectMaybeShared {
       // `std::malloc` returns memory at least as strictly aligned as for
       // max_align_t and the alignment of max_align_t is a multiple of the array
       // buffer alignment.
-      static_assert(alignof(std::max_align_t) %
+      static_assert(JS_MALLOC_ALIGNMENT %
                         ArrayBufferObject::ARRAY_BUFFER_ALIGNMENT ==
                     0);
 
diff --git a/js/src/vm/TypedArrayObject-inl.h b/js/src/vm/TypedArrayObject-inl.h
index 2702836452..65bd4619de 100644
--- a/js/src/vm/TypedArrayObject-inl.h
+++ b/js/src/vm/TypedArrayObject-inl.h
@@ -385,7 +385,7 @@ class ElementSpecific {
     // `malloc` returns memory at least as strictly aligned as for max_align_t
     // and the alignment of max_align_t is a multiple of the size of `T`,
     // so `SharedMem::cast` will be called with properly aligned memory.
-    static_assert(alignof(std::max_align_t) % sizeof(T) == 0);
+    static_assert(JS_MALLOC_ALIGNMENT % sizeof(T) == 0);
 
     SharedMem<T*> dest = Ops::extract(target).template cast<T*>() + offset;
     SharedMem<void*> data = Ops::extract(source);
diff --git a/js/src/wasm/WasmSignalHandlers.cpp b/js/src/wasm/WasmSignalHandlers.cpp
index 817f93cc6c..761e9be734 100644
--- a/js/src/wasm/WasmSignalHandlers.cpp
+++ b/js/src/wasm/WasmSignalHandlers.cpp
@@ -369,7 +369,8 @@ struct macos_x86_context {
 #    elif defined(__arm__)
 struct macos_arm_context {
   arm_thread_state_t thread;
-  arm_neon_state_t float_;
+  // armv7 iOS has no ARM_NEON_STATE; the VFP state holds d0-d31
+  arm_vfp_state_t float_;
 };
 #      define CONTEXT macos_arm_context
 #    elif defined(__aarch64__)
@@ -663,9 +664,9 @@ static bool HandleMachException(const ExceptionRequest& request) {
   int float_state = x86_FLOAT_STATE;
 #    elif defined(__arm__)
   unsigned int thread_state_count = ARM_THREAD_STATE_COUNT;
-  unsigned int float_state_count = ARM_NEON_STATE_COUNT;
+  unsigned int float_state_count = ARM_VFP_STATE_COUNT;
   int thread_state = ARM_THREAD_STATE;
-  int float_state = ARM_NEON_STATE;
+  int float_state = ARM_VFP_STATE;
 #    elif defined(__aarch64__)
   unsigned int thread_state_count = ARM_THREAD_STATE64_COUNT;
   unsigned int float_state_count = ARM_NEON_STATE64_COUNT;
diff --git a/memory/build/Mutex.cpp b/memory/build/Mutex.cpp
index bbdf7fa2be..1cf492ff37 100644
--- a/memory/build/Mutex.cpp
+++ b/memory/build/Mutex.cpp
@@ -13,7 +13,7 @@ bool Mutex::TryLock() {
 
 #if defined(XP_WIN)
   return !!TryEnterCriticalSection(mMutex.addr());
-#elif defined(XP_DARWIN)
+#elif defined(MOZ_MUTEX_UNFAIR_LOCK)
   return os_unfair_lock_trylock(&mMutex);
 #else
   switch (pthread_mutex_trylock(&mMutex)) {
diff --git a/memory/build/Mutex.h b/memory/build/Mutex.h
index a9fa6b549e..9d1d775e05 100644
--- a/memory/build/Mutex.h
+++ b/memory/build/Mutex.h
@@ -5,12 +5,18 @@
 #ifndef Mutex_h
 #define Mutex_h
 
+// os_unfair_lock came with iOS 10; armv7 iOS (LegacyFox, iOS 4-7) takes the
+// pthread mutex like the other Unixes.
+#if defined(XP_DARWIN) && !(defined(XP_IOS) && defined(__arm__))
+#  define MOZ_MUTEX_UNFAIR_LOCK 1
+#endif
+
 #if defined(XP_WIN)
 #  include <windows.h>
 #else
 #  include <pthread.h>
 #endif
-#if defined(XP_DARWIN)
+#if defined(MOZ_MUTEX_UNFAIR_LOCK)
 #  include <os/lock.h>
 #endif
 
@@ -19,7 +25,7 @@
 #include "mozilla/MaybeStorageBase.h"
 #include "mozilla/ThreadSafety.h"
 
-#if defined(XP_DARWIN)
+#if defined(MOZ_MUTEX_UNFAIR_LOCK)
 // For information about the following undocumented flags and functions see
 // https://github.com/apple/darwin-xnu/blob/main/bsd/sys/ulock.h and
 // https://github.com/apple/darwin-libplatform/blob/main/private/os/lock_private.h
@@ -33,7 +39,7 @@ OS_UNFAIR_LOCK_AVAILABILITY
 OS_EXPORT OS_NOTHROW OS_NONNULL_ALL void os_unfair_lock_lock_with_options(
     os_unfair_lock_t lock, os_unfair_lock_options_t options);
 }
-#endif  // defined(XP_DARWIN)
+#endif  // defined(MOZ_MUTEX_UNFAIR_LOCK)
 
 // Mutexes are based on spinlocks.  We can't use normal pthread spinlocks in all
 // places, because they require malloc()ed memory, which causes bootstrapping
@@ -48,7 +54,7 @@ struct MOZ_CAPABILITY("mutex") Mutex {
 #if defined(XP_WIN)
   // MaybeStorageBase provides a constexpr constructor.
   mozilla::detail::MaybeStorageBase<CRITICAL_SECTION> mMutex;
-#elif defined(XP_DARWIN)
+#elif defined(MOZ_MUTEX_UNFAIR_LOCK)
   os_unfair_lock mMutex = OS_UNFAIR_LOCK_INIT;
 #elif defined(XP_LINUX) && !defined(ANDROID)
   pthread_mutex_t mMutex = PTHREAD_ADAPTIVE_MUTEX_INITIALIZER_NP;
@@ -78,7 +84,7 @@ struct MOZ_CAPABILITY("mutex") Mutex {
     if (!InitializeCriticalSectionAndSpinCount(mMutex.addr(), 5000)) {
       return false;
     }
-#elif defined(XP_DARWIN)
+#elif defined(MOZ_MUTEX_UNFAIR_LOCK)
     mMutex = OS_UNFAIR_LOCK_INIT;
 #elif defined(XP_LINUX) && !defined(ANDROID)
     pthread_mutexattr_t attr;
@@ -104,7 +110,7 @@ struct MOZ_CAPABILITY("mutex") Mutex {
 
 #if defined(XP_WIN)
     EnterCriticalSection(mMutex.addr());
-#elif defined(XP_DARWIN)
+#elif defined(MOZ_MUTEX_UNFAIR_LOCK)
     // We rely on a non-public function to improve performance here.
     // The OS_UNFAIR_LOCK_DATA_SYNCHRONIZATION flag informs the kernel that
     // the calling thread is able to make progress even in absence of actions
@@ -127,14 +133,14 @@ struct MOZ_CAPABILITY("mutex") Mutex {
 
 #if defined(XP_WIN)
     LeaveCriticalSection(mMutex.addr());
-#elif defined(XP_DARWIN)
+#elif defined(MOZ_MUTEX_UNFAIR_LOCK)
     os_unfair_lock_unlock(&mMutex);
 #else
     pthread_mutex_unlock(&mMutex);
 #endif
   }
 
-#if defined(XP_DARWIN)
+#if defined(MOZ_MUTEX_UNFAIR_LOCK)
   static bool SpinInKernelSpace();
   static const bool gSpinInKernelSpace;
 #endif  // XP_DARWIN
-- 
2.53.0

